### PR TITLE
159 update api addresses

### PR DESCRIPTION
**Pull request by markWarr**
Update API to point to Cosmos DB, add and obscure secrets.

### CODE REVIEW

**Review by agnieszkazbikowska-madetech — DISMISSED** [review]

File: HousingManagementSystemApi.Tests/GatewaysTests/AddressesCosmosGatewayTests.cs [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+                .Returns(this.feedIteratorMock.Object);
+
+            this.cosmosQueryHelperMock.Setup(_ => _.GetItemQueryIterator<PropertyAddress>("NG21 9LQ"))
+                .Returns(this.feedIteratorMock.Object);
+
+            // Act
+            var results = await this.systemUnderTest.SearchByPostcode("NG21 9LQ");
+
+            // Assert
+            Assert.True(results.Count() == 1);
```
Comment (Bhavesh-Patel) [inline HousingManagementSystemApi.Tests/GatewaysTests/AddressesCosmosGatewayTests.cs:94]: This assert wouldn't show an informative message on failure:
```
Expected: True
Actual:   False
```

An informative error message can be useful in understanding the reason for failure.
Using `Assert.Equal(1, results.Counts())` would give the following error message:
```
Expected: 1
Actual:   0
```
Using `Assert.Single(results)` gives the following error message:
```
The collection was expected to contain a single element, but it was empty.
```

The latter example is preferred, i.e.
```suggestion
            Assert.Single(results);
```

File: HousingManagementSystemApi.Tests/HelpersTests/CosmosAddressQueryHelperTests.cs [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+        public CosmosAddressQueryHelperTests()
+        {
+            this.containerMock = new Mock<Container>();
+            this.systemUnderTest = new CosmosAddressQueryHelper(this.containerMock.Object);
+            this.feedIteratorMock = new Mock<FeedIterator<PropertyAddress>>();
+        }
+
+        [Fact]
+        public void Test_Query_Definition_returned_with_querytext()
```
Comment (Bhavesh-Patel) [inline HousingManagementSystemApi.Tests/HelpersTests/CosmosAddressQueryHelperTests.cs:23]: The underscores in the test names are raised as warnings during the build due to breaking defined naming conventions - but in the case of test names it was decided to allow it.
Add a disable pragma to remove the warnings, i.e.
```suggestion
#pragma warning disable CA1707
        public void Test_Query_Definition_returned_with_querytext()
#pragma warning restore CA1707
```

The comment applies to each of the tests in this PR.

Comment (Bhavesh-Patel) [inline HousingManagementSystemApi.Tests/HelpersTests/CosmosAddressQueryHelperTests.cs:23]: Correct test name casing
```suggestion
        public void Test_Query_Definition_Returned_With_QueryText()
```

File: HousingManagementSystemApi.Tests/HelpersTests/CosmosAddressQueryHelperTests.cs [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        {
+            // Arrange
+            this.feedIteratorMock.Setup(_ => _.HasMoreResults).Returns(false);
+            this.containerMock
+                .Setup(_ => _.GetItemQueryIterator<PropertyAddress>("TEST", null, null)) //It.IsAny<string>()
```
Comment (Bhavesh-Patel) [inline HousingManagementSystemApi.Tests/HelpersTests/CosmosAddressQueryHelperTests.cs:28]: Remove commented code
```suggestion
                .Setup(_ => _.GetItemQueryIterator<PropertyAddress>("TEST", null, null))
```

File: HousingManagementSystemApi.Tests/HelpersTests/CosmosAddressQueryHelperTests.cs [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+                .Returns(this.feedIteratorMock.Object);
+
+            // Act
+            var result = this.systemUnderTest.GetItemQueryIterator<PropertyAddress>("NG21 7HT");
```
Comment (Bhavesh-Patel) [inline HousingManagementSystemApi.Tests/HelpersTests/CosmosAddressQueryHelperTests.cs:32]: The result is not used, remove assignment to make this clear.
```suggestion
            this.systemUnderTest.GetItemQueryIterator<PropertyAddress>("NG21 7HT");
```

File: HousingManagementSystemApi.Tests/HelpersTests/CosmosAddressQueryHelperTests.cs [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+            // Assert
+            this.containerMock.Verify(m => m.GetItemQueryIterator<PropertyAddress>(
+                It.Is<QueryDefinition>(u =>
+                        u.QueryText ==
+                        "SELECT * FROM c WHERE (UPPER(REPLACE(c.PostalCode, ' ','')))  = (UPPER(REPLACE(@postcode, ' ','')))  ORDER BY c.AddressLine[0] ASC"
```
Comment (Bhavesh-Patel) [inline HousingManagementSystemApi.Tests/HelpersTests/CosmosAddressQueryHelperTests.cs:38]: Is this test verifying that the correct SQL is used?

File: HousingManagementSystemApi.Tests/HelpersTests/CosmosAddressQueryHelperTests.cs [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+                , It.IsAny<string>(), null));
+        }
+
+        [Fact]
+        public void Test_Query_Definition_returned_with_parameter()
+        {
+            // Arrange
+            this.feedIteratorMock.Setup(_ => _.HasMoreResults).Returns(false);
+            this.containerMock
+                .Setup(_ => _.GetItemQueryIterator<PropertyAddress>("TEST", null, null)) //It.IsAny<string>()
```
Comment (Bhavesh-Patel) [inline HousingManagementSystemApi.Tests/HelpersTests/CosmosAddressQueryHelperTests.cs:49]: Remove commented code
```suggestion
                .Setup(_ => _.GetItemQueryIterator<PropertyAddress>("TEST", null, null))
```

File: HousingManagementSystemApi.Tests/HelpersTests/CosmosAddressQueryHelperTests.cs [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+            this.containerMock.Verify(m => m.GetItemQueryIterator<PropertyAddress>(
+                It.Is<QueryDefinition>(u =>
+                        u.QueryText ==
+                        "SELECT * FROM c WHERE (UPPER(REPLACE(c.PostalCode, ' ','')))  = (UPPER(REPLACE(@postcode, ' ','')))  ORDER BY c.AddressLine[0] ASC"
+                )
+                , It.IsAny<string>(), null));
+        }
+
+        [Fact]
+        public void Test_Query_Definition_returned_with_parameter()
```
Comment (Bhavesh-Patel) [inline HousingManagementSystemApi.Tests/HelpersTests/CosmosAddressQueryHelperTests.cs:44]: Correct test name casing
```suggestion
        public void Test_Query_Definition_Returned_With_Parameter()
```

File: HousingManagementSystemApi/Gateways/AddressesCosmosGateway.cs [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+
+        public async Task<IEnumerable<PropertyAddress>> SearchByPostcode(string postcode)
+        {
+            using var queryResultSetIterator = this.cosmosAddressQueryHelper.GetItemQueryIterator<PropertyAddress>(postcode);
+            var addresses = new List<PropertyAddress>();
+
+            while (queryResultSetIterator.HasMoreResults)
+            {
+                var currentResultSet = await queryResultSetIterator.ReadNextAsync();
+                foreach (var address in currentResultSet)
```
Comment (Bhavesh-Patel) [inline HousingManagementSystemApi/Gateways/AddressesCosmosGateway.cs:25]: Replace use of `foreach` loop with single method call, i.e.
```suggestion
                addresses.AddRange(currentResultSet);
```

File: HousingManagementSystemApi/Startup.cs [hunk 9]
```diff
@@ -76,6 +81,15 @@ public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
...
+        private static Container getCosmosClientContainer()
+        {
+            var cosmosClient = new CosmosClient(EnvironmentVariableHelper.GetEnvironmentVariable("COSMOS_ENDPOINT_URL"),
+                EnvironmentVariableHelper.GetEnvironmentVariable("COSMOS_AUTHORIZATION_KEY"));
+
+            return cosmosClient.GetContainer(EnvironmentVariableHelper.GetEnvironmentVariable("COSMOS_DATABASE_ID"),
+                EnvironmentVariableHelper.GetEnvironmentVariable("COSMOS_TENANT_CONTAINER_ID"));
+        }
+
         private static string GetEnvironmentVariable(string name) =>
```
Comment (Bhavesh-Patel) [inline HousingManagementSystemApi/Startup.cs:93]: This method has been relocated to `EnvironmentVariableHelper` ([here](https://github.com/Newarkandsherwood/HousingManagementSystemApi/pull/13/files#diff-b13b4fcdbabb3b91a4c525d69232225e638dce5c2832dcafee99356faf1811ea)) and should be deleted from here as this no longer  used and is code duplication.

**Review by Bhavesh-Patel — CHANGES_REQUESTED** [review]

File: HousingManagementSystemApi.Tests/GatewaysTests/AddressesCosmosGatewayTests.cs [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+            this.systemUnderTest = new AddressesCosmosGateway(this.cosmosQueryHelperMock.Object);
+            this.feedIteratorMock = new Mock<FeedIterator<PropertyAddress>>();
+        }
+
+        [Fact]
+        public async void Test_Get_addresses_Not_Throws_Exception()
+        {
+            // Arrange
+            this.feedIteratorMock.Setup(_ => _.HasMoreResults).Returns(false);
+            this.cosmosQueryHelperMock.Setup(_ => _.GetItemQueryIterator<PropertyAddress>("NG21 9LQ"))
```
[the remark this answers is not in the document]
Comment (agnieszkazbikowska-madetech) [reply HousingManagementSystemApi.Tests/GatewaysTests/AddressesCosmosGatewayTests.cs]: It makes sense. I've created a `constant`.

File: HousingManagementSystemApi.Tests/GatewaysTests/AddressesCosmosGatewayTests.cs [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+            this.cosmosQueryHelperMock = new Mock<ICosmosAddressQueryHelper>();
+            this.systemUnderTest = new AddressesCosmosGateway(this.cosmosQueryHelperMock.Object);
+            this.feedIteratorMock = new Mock<FeedIterator<PropertyAddress>>();
+        }
+
+        [Fact]
+        public async void Test_Get_addresses_Not_Throws_Exception()
+        {
+            // Arrange
+            this.feedIteratorMock.Setup(_ => _.HasMoreResults).Returns(false);
```
[the remark this answers is not in the document]
Comment (markWarr) [reply HousingManagementSystemApi.Tests/GatewaysTests/AddressesCosmosGatewayTests.cs]: These have been removed and .editorconfig updated.

File: HousingManagementSystemApi.Tests/GatewaysTests/AddressesCosmosGatewayTests.cs [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+            // Assert
+            Assert.True(!results.Any());
+        }
+
+        [Fact]
+        public async void Test_Addresses_Returned_When_Iterator_Results()
+        {
+            // Arrange
+            this.feedIteratorMock.Setup(_ => _.HasMoreResults).Returns(true);
+            var addressList = new List<PropertyAddress>();
```
[the remark this answers is not in the document]
Comment (markWarr) [reply HousingManagementSystemApi.Tests/GatewaysTests/AddressesCosmosGatewayTests.cs]: Fix committed.

File: HousingManagementSystemApi.Tests/HelpersTests/CosmosAddressQueryHelperTests.cs (hunk 5 above)
Comment (markWarr) [reply HousingManagementSystemApi.Tests/HelpersTests/CosmosAddressQueryHelperTests.cs:39]: We are testing that the Querytext is set, and that the SQL it is set to is correct.

File: HousingManagementSystemApi.Tests/GatewaysTests/AddressesCosmosGatewayTests.cs (hunk 1 above)
Comment (markWarr) [reply HousingManagementSystemApi.Tests/GatewaysTests/AddressesCosmosGatewayTests.cs:93]: Thanks, I didn't know about this.

**Review by agnieszkazbikowska-madetech — DISMISSED** [review]

File: HousingManagementSystemApi.Tests/ContollersTests/AddressControllerTests.cs [hunk 13]
```diff
@@ -71,7 +72,7 @@ public async Task GivenAnExceptionIsThrown_WhenRequestMadeForAddresses_ResponseH
             // Arrange
             const string errorMessage = "An error message";
             retrieveAddressesUseCaseMock.Setup(x => x.Execute(It.IsAny<string>()))
-                .Throws(new Exception(errorMessage));
+                .Throws(new ConfigurationErrorsException(errorMessage));
```
Comment (Bhavesh-Patel) [inline HousingManagementSystemApi.Tests/ContollersTests/AddressControllerTests.cs:75]: Is this really a configuration error?
Given the system under test, AddresssController, catches `Exception` ([here](https://github.com/Newarkandsherwood/HousingManagementSystemApi/blob/main/HousingManagementSystemApi/Controllers/AddressesController.cs#L28)), I feel it's appropriate to use the same exception here.
This test doesn't cover the scenario where the use case throws a different exception and ensures that exception is handling by AddressController.

Given this warning existed prior to any of these changes and no changes to the AddressController were made that resulted in the warning, I suggest restoring the test to use `Exception`.

The (pre-existing) CI warning should be tackled independently.

`ConfigurationErrorsException` would be ideal to use when our configuration isn't setup correctly, e.g. [EnvironmentVariableHelper](https://github.com/Newarkandsherwood/housing-repairs-online-api/blob/09d74ab082b600bce8eaab26fcd006c58abeeb63/HousingRepairsOnlineApi/Helpers/EnvironmentVariableHelper.cs#L10)

Comment (markWarr) [reply HousingManagementSystemApi.Tests/ContollersTests/AddressControllerTests.cs]: I was going through trying to get rid of warnings. You're right, though - this was pre existing and whilst this change got rid of the warning, it's best left for now.

**Review by Bhavesh-Patel — APPROVED** [review]

**Pull request merged**